hphp/runtime/vm/jit/relocation-ppc64.cpp

```diff
@@ -0,0 +1,595 @@
+    }
+  }
+
+  decltype(meta.addressImmediates) updatedAI;
+  for (auto addrImm : meta.addressImmediates) {
+    if (TCA adjusted = rel.adjustedAddressAfter(addrImm)) {
+      updatedAI.insert(adjusted);
+    } else if (TCA odd = rel.adjustedAddressAfter((TCA)~uintptr_t(addrImm))) {
+      // just for cgLdObjMethod
+      updatedAI.insert((TCA)~uintptr_t(odd));
+    } else {
+      updatedAI.insert(addrImm);
+    }
+  }
+  updatedAI.swap(meta.addressImmediates);
+
+  decltype(meta.alignments) updatedAF;
+  for (auto af : meta.alignments) {
+    if (TCA adjusted = rel.adjustedAddressAfter(af.first)) {
+      updatedAF.emplace(adjusted, af.second);
+    } else {
+      updatedAF.emplace(af);
+    }
+  }
+  updatedAF.swap(meta.alignments);
+
+  for (auto& af : meta.reusedStubs) {
+    if (TCA adjusted = rel.adjustedAddressAfter(af)) {
+      af = adjusted;
+    }
+  }
+
+  if (asmInfo) {
+    assert(asmInfo->validate());
+    fixupRanges(asmInfo, AreaIndex::Main, rel);
+    fixupRanges(asmInfo, AreaIndex::Cold, rel);
+    fixupRanges(asmInfo, AreaIndex::Frozen, rel);
+    assert(asmInfo->validate());
+  }
+}
+
+/*
+ * Adjust potentially live references that point into the relocated
+ * area.
+ * Must not be called until its safe to run the relocated code.
+ */
+void adjustCodeForRelocation(RelocationInfo& rel, CGMeta& fixups) {
+  for (auto addr : fixups.reusedStubs) {
+    /*
+     * The stubs are terminated by a trap. Check for it.
+     */
+    for (auto di = DecodedInstruction(addr); !di.isException(); ) {
+      adjustInstruction(rel, di);
+
+      addr += di.size();
+      di = DecodedInstruction(addr);
+    }
+  }
+
+  for (auto codePtr : fixups.codePointers) {
+    if (TCA adjusted = rel.adjustedAddressAfter(*codePtr)) {
+      *codePtr = adjusted;
+    }
+  }
+}
+
+void findFixups(TCA start, TCA end, CGMeta& meta) {
+  while (start != end) {
+    assert(start < end);
+    DecodedInstruction di(start);
+    start += di.size();
+
+    if (di.isCall()) {
+      if (auto fixup = FixupMap::findFixup(start)) {
+        meta.fixups.emplace_back(start, *fixup);
+      }
+      if (auto ct = getCatchTrace(start)) {
+        meta.catches.emplace_back(start, *ct);
+      }
+    }
+  }
+}
+
+
+/*
+ * Relocate code in the range start, end into dest, and record
+ * information about what was done to rel.
+ * On exit, internal references (references into the source range)
+ * will have been adjusted (ie they are still references into the
+ * relocated code). External code references continue to point to
+ * the same address as before relocation.
+ */
+size_t relocate(RelocationInfo& rel,
+                CodeBlock& dest_block,
+                TCA start, TCA end,
+                CGMeta& fixups,
+                TCA* exit_addr) {
+  JmpSet far_jmps, near_jmps;
+  while (true) {
+    try {
+      return relocateImpl(rel, dest_block, start, end,
+                          fixups, exit_addr, far_jmps, near_jmps);
+    } catch (JmpOutOfRange& j) {
+    }
```

Comment:
No, just like X64: it's suppose to retry and it's guaranteed that it will progress on the next iteration as it only throws if new data is stored on a `JmpSet` data. That new data will make it possible to re-relocate further ahead.